crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -229,6 +258,7 @@ impl<'a> Checker<'a> {
     #[allow(clippy::too_many_arguments)]
     pub(crate) fn new(
         parsed: &'a Parsed<ModModule>,
+        parsed_annotations_arena: &'a typed_arena::Arena<AnnotationParseResult>,
```

Comment:
I just pushed https://github.com/astral-sh/ruff/pull/13158/commits/8ff61ad8f16aa9b88f45f653b170acd711769aac to clarify the design of the cache a little bit.

The `HashMap` in the cache has to be behind a `RefCell` or we can't call `Checker::parse_type_annotation` with an `&self` reference, and having it take an `&mut self` reference causes all sorts of borrow checker complaints. If the `HashMap` behind the `RefCell` owns the data, however, it's impossible to return a reference to the data from the cache's `lookup_or_parse()` method, or the borrow checker complains that the data returned by `lookup_or_parse()` borrows data owned by the current function. So the parsed annotations have to be stored in some kind of external storage outside of the `RefCell<HashMap>`, with the `RefCell<HashMap>` only storing references to the data.

If there's a better way of doing this, I'd love to know about it. But I've been puzzling over this for much of the afternoon, and can't really see it :(

Reply:
I don't think there's a better way. 

As you said, the value returned by `lookup_or_parse` must be copy to avoid borrowing a value that's behind a `RefCell`. 

This is accomplished by storing references that outlive the `Cache` (and `Checker`))